File: notification_system_impl_python/providers/fcm_provider.py
import json
import os
from typing import Dict, Any, Optional
from providers.base_provider import NotificationProvider
import requests
from dotenv import load_dotenv

load_dotenv()


class FCMProvider(NotificationProvider):
    def __init__(self, server_key: str):
        self.server_key = server_key
        self.api_url = "https://fcm.googleapis.com/fcm/send"

    def provider_name(self) -> str:
        return "fcm"

    def send(self, notification) -> Dict[str, Any]:
        try:
            payload = json.loads(notification.payload)
            token = payload.get('token')
            topic = payload.get('topic')

            if not token and not topic:
                return {'success': False, 'message': 'Missing "token" or "topic" field in payload'}
            fcm_message = {
                'notification': {
                    'title': payload.get('title', 'Notification'),
                    'body': payload.get('body', '')
                }
            }
            if 'data' in payload:
                fcm_message['data'] = payload['data']

            if token:
                fcm_message['to'] = token
            else:
                fcm_message['to'] = f'/topics/{topic}'

            headers = {
                'Content-Type': 'application/json',
                'Authorization': f'key={self.server_key}'
            }

            response = requests.post(
                self.api_url,
                headers=headers,
                json=fcm_message,
                timeout=10
            )
            if response.status_code >= 400:
                return {
                    'success': False,
                    'message': f'FCM returned error status {response.status_code}: {response.text}'
                }

            result = response.json()
            success_count = result.get('success', 0)
            failure_count = result.get('failure', 0)

            if success_count > 0:
                return {
                    'success': True,
                    'message': 'Push notification sent via FCM',
                    'response': result
                }
            error_msg = 'Unknown error'
            if failure_count > 0 and 'results' in result:
                results = result['results']
                if results and 'error' in results[0]:
                    error_msg = results[0]['error']

            return {
                'success': False,
                'message': f'FCM error: {error_msg}',
                'response': result
            }

        except requests.RequestException as e:
            return {
                'success': False,
                'message': f'FCM request failed: {str(e)}'
            }
        except Exception as e:
            return {
                'success': False,
                'message': f'FCM send failed: {str(e)}'
            }


def load_fcm_provider_from_env() -> Optional[FCMProvider]:
    server_key = os.getenv("FCM_SERVER_KEY")
    if server_key:
        return FCMProvider(server_key)
    return None

File: notification_system_impl_go/core/providers/fcm_provider.go
package providers

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"notification_system/models"
)

type FCMProvider struct {
	ServerKey string
}

func NewFCMProvider(serverKey string) *FCMProvider {
	return &FCMProvider{
		ServerKey: serverKey,
	}
}

func (f *FCMProvider) ProviderName() string {
	return "fcm"
}

func (f *FCMProvider) Send(ctx context.Context, notification *models.Notification) error {
	var payload map[string]interface{}
	if err := json.Unmarshal([]byte(notification.Payload), &payload); err != nil {
		return fmt.Errorf("invalid payload: %w", err)
	}

	token, hasToken := payload["token"].(string)
	topic, hasTopic := payload["topic"].(string)

	if !hasToken && !hasTopic {
		return fmt.Errorf("missing 'token' or 'topic' field in payload")
	}

	fcmMessage := map[string]interface{}{
		"notification": map[string]interface{}{
			"title": payload["title"],
			"body":  payload["body"],
		},
	}

	if data, ok := payload["data"].(map[string]interface{}); ok {
		fcmMessage["data"] = data
	}

	if hasToken {
		fcmMessage["to"] = token
	} else {
		fcmMessage["to"] = "/topics/" + topic
	}

	jsonData, err := json.Marshal(fcmMessage)
	if err != nil {
		return fmt.Errorf("failed to marshal FCM message: %w", err)
	}

	req, err := http.NewRequestWithContext(
		ctx,
		"POST",
		"https://fcm.googleapis.com/fcm/send",
		bytes.NewBuffer(jsonData),
	)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "key="+f.ServerKey)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("failed to send FCM notification: %w", err)
	}
	defer resp.Body.Close()

	responseBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode >= 400 {
		return fmt.Errorf("FCM returned error status %d: %s", resp.StatusCode, string(responseBody))
	}

	var result map[string]interface{}
	if err := json.Unmarshal(responseBody, &result); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	if success, ok := result["success"].(float64); ok && success > 0 {
		notification.Status = models.StatusSent
		now := time.Now().UnixMilli()
		notification.SentAt = &now
		responseStr := string(responseBody)
		notification.ProviderResponse = &responseStr
		return nil
	}

	if failure, ok := result["failure"].(float64); ok && failure > 0 {
		errorMsg := "FCM send failed"
		if results, ok := result["results"].([]interface{}); ok && len(results) > 0 {
			if firstResult, ok := results[0].(map[string]interface{}); ok {
				if err, ok := firstResult["error"].(string); ok {
					errorMsg = err
				}
			}
		}
		return fmt.Errorf("FCM error: %s", errorMsg)
	}

	return fmt.Errorf("unexpected FCM response: %s", string(responseBody))
}
